[#7288][feat] Added AutoDeploy backend support to test_perf.py

MrGeva opened the pull request:
Added support for AD in test_perf.py. for AD it fails the test when perf regression detected. On PT it waives the perf regression as it was done so far - no change for PT test behavior.
Removed the old AD perf test which was disabled anyway, and removed it from the yaml files.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Performance tests now support an autodeploy backend, selectable in configs and CLI strings.
  * Added configuration options with defaults: compile backend, free memory ratio, runtime, and an option to skip loading weights.
  * Bench commands and metrics treat autodeploy like PyTorch, skipping build steps and build-time metrics where applicable.
* **Tests**
  * Updated parameter parsing and test naming to recognize and display the autodeploy backend.
  * Removed the old AD perf test: trtllm_bench_backend_comparison
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

MrGeva commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17931](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17931/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #13437](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13437/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #17939](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17939/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13444](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13444/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/13444/test-results/rerun_report.html) for details.

tensorrt-cicd commented:
[PR_Github #17982](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17982/) [ run ] triggered by Bot

MrGeva commented:
/bot run

tensorrt-cicd commented:
[PR_Github #18016](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18016/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18047](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18047/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13523 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13523/) completed with status: 'SUCCESS'

MrGeva commented:
/bot run --disable-reuse-test

tensorrt-cicd commented:
[PR_Github #18054](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18054/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18187](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18187/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18187](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18187/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13630](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13630/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #18364](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18364/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18400](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18400/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18400](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18400/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13802](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13802/) completed with status: 'FAILURE'

MrGeva commented:
/bot run

MrGeva commented:
/bot run

tensorrt-cicd commented:
[PR_Github #18745](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18745/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18745](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18745/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #14057](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14057/) completed with status: 'FAILURE'

MrGeva commented on tests/integration/defs/perf/test_perf.py:
```diff
@@ -1299,6 +1313,25 @@ def get_trtllm_bench_command(self, engine_dir):
+            with open(autodeploy_config_path, 'w') as f:
+                yaml.dump(autodeploy_config, f, default_flow_style=False)
+            benchmark_cmd += [f"--extra_llm_api_options={autodeploy_config_path}"]
         return benchmark_cmd
```
_get_engine_dir generates a unique folder name according to the test params

MrGeva commented:
/bot run

tensorrt-cicd commented:
[PR_Github #18795](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18795/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14089/" target="_blank">LLM/main/L0_MergeRequest_PR #14089</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14089//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

tensorrt-cicd commented:
[PR_Github #18800](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18800/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14094](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14094/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #19022](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19022/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19021](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19021/) [ run ] completed with state `ABORTED`

suyoggupta commented on tests/integration/defs/perf/test_perf.py:
```diff
@@ -408,6 +414,11 @@ def __init__(
+        ad_compile_backend: str = "torch-opt",
+        free_mem_ratio: float = 0.9,
+        extra_runtime: str = "trtllm",
+        skip_loading_weights: bool = False,
```
should we explicitly specify cudagraph batch sizes here?

tensorrt-cicd commented:
[PR_Github #19190](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19190/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19190](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19190/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14403](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14403/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #19195](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19195/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19195](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19195/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #14408](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14408/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #19208](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19208/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19322](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19322/) [ run ] triggered by Bot

MrGeva commented:
/bot run

tensorrt-cicd commented:
[PR_Github #19435](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19435/) [ run ] triggered by Bot

omera-nv commented on jenkins/L0_Test.groovy:
```diff
@@ -1765,11 +1765,17 @@ def runLLMTestlistOnPlatformImpl(pipeline, platform, testList, config=VANILLA_CO
+                    """,
+                    returnStatus: true
+                )
+                if (perfCheckResult != 0) {
```
Was this discussed somewhere? I think doing such a drastic change as failing pipelines if a perf regression is detected should be done in a dedicated PR, and broadcasted clearly to all contributors.

omera-nv commented on tests/integration/test_lists/test-db/l0_perf.yml:
```diff
@@ -29,3 +29,17 @@ l0_perf:
+        stage: pre_merge
+        backend: pytorch
+    tests:
+      - perf/test_perf.py::test_perf[llama_v3.1_8b_instruct-bench-_autodeploy-float16-input_output_len:128,128-reqs:8192]
```
I wanted to check how long this new test takes but the latest pipeline already expired in Jenkins, do you have an estimate of how much it takes to complete?

omera-nv reviewed (commented):
I only reviewed the parts related to testing infra, head two questions

MrGeva committed:
Add AD backend support for performance testing

- Added AD backend support in test_perf.py
- Added test to l0_perf.yml with proper configuration
- Updated performance thresholds and regression handling
- Added KV cache metrics and threshold support
- Simplified regex handling for KV cache
- Removed old AD perf test and trtllm_bench_backend_comparison
- Fixed import coding style and various path/param issues

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

MrGeva committed:
fixed wrong rebase

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #20178](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20178/) [ run ] triggered by Bot

MrGeva commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20182](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20182/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20182](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20182/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15219](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15219/) completed with status: 'SUCCESS'
